### ISSUE TITLE
Is the objectness score really necessary?

### ISSUE THREAD

**Issue by mmartin56** [AlexeyAB/darknet#4042]
Hi @AlexeyAB,

I'm curious about the usefulness of Yolo's objectness score and I hope you can shed some light on this. Yolo gives two groups of scores for each bounding box:
**1.the objectness score (scalar in [0,1])
2. the class score, one for each class (array of scalars in [0,1])**
They are basically yolo's answer to two questions:
**1. What's your confidence that this bounding box is wrapped around an object accurately?
2. If this bounding box was indeed an object, what's your confidence that this object is in each particular class?**

I was wondering, for the sake of knowledge, what would be the downside of getting rid of the objectness score altogether? And have class scores also implicitly carry objectness scores, to basically answer the question
**What's your confidence that this bounding box is in an object of each particular class?**

The code would need to be changed a bit: delta scores in yolo_layer.c in particular, and in a few other places.

Basically I don't really see the point of this two-stage scoring instead of a more direct scoring.

**Comment by AlexeyAB** [comment]
@AlexeyAB Hi,

It seems that `objectness` solves the same issue as `Focal loss`, i.e. it solves Class disbalance issue - when there are objects with class_0 much more than objects with class_1 in the training dataset. So both classes will compete only for Probability, but not for Objectness.

https://arxiv.org/pdf/1804.02767v1.pdf

> Focal loss. We tried using focal loss. It dropped our
> mAP about 2 points. YOLOv3 may already be robust to
> the problem focal loss is trying to solve because it has separate objectness predictions and conditional class predictions. Thus for most examples there is no loss from the
> class predictions? Or something? We aren’t totally sure.

----

But there are many ambiguities: 
1. why we should use 
    `probability = objectness * class_probability` and `class_id = max(probability[i])`
     instead of:
    `probability = objectness` and `class_id = max(class_probability[i])`

2. why we take into account only `IoU(pred, truth)`, but doesn't take into account `class_id` match there: https://github.com/AlexeyAB/darknet/blob/53160fa6662ea7e8c4095588db56cac7b339b917/src/yolo_layer.c#L294

3. why should we use `ignore_thresh` rather than `truth_thresh`  https://github.com/AlexeyAB/darknet/blob/53160fa6662ea7e8c4095588db56cac7b339b917/src/yolo_layer.c#L291-L306


https://github.com/AlexeyAB/darknet/blob/53160fa6662ea7e8c4095588db56cac7b339b917/cfg/yolov3.cfg#L786-L787
Why it is better than `ignore_thresh = .7  truth_thresh = 0.7` ?

**Comment by mmartin56** [comment]
Hi Alexey,

Thanks for your answer. 

1) The reason is that it seems to me like a more natural and direct approach. The current approach is somewhat indirect (objectness and class scores vs just class score). I don't see what objectness achieves.

2) I don't understand your point here. `IoU(pred, truth)` and `l.ignore_thresh` are unrelated to objectness and class scores, so that doesn't need to change. It's basically saying "do not change objectness nor class scores if the bounding box overlaps too much with a ground truth object".

3) I don't understand your point here either. As I questioned [there](https://github.com/AlexeyAB/darknet/issues/4015), `l.truth_thresh = 1` in all cfg files. Since `best_iou` <= 1 (always), this if statement is never satisfied, so lines 298-305 are never run. So nothing to change there. [The 'match' between ground truth objects and 'responsible' predictors - which leads to an increase, for those predictors, in objectness and in the object's class score - is treated in subsequent lines in yolo_layer.c]

I might have misunderstood your points though. Could you please clarify what you mean?

Finally on your point on class imbalance: say there are 100 aeroplanes and 900 persons in your dataset. Whether we follow the current approach or the approach I mentioned, the ratio between the 'aeroplane' and 'person' class scores should be the same in both approaches: during training, when a predictor is responsible for a ground truth object, the 'aeroplane' class score increases on average 10% of the time, and decreases 90% of the time (and inversely for the 'person' class score).

So, wouldn't this new approach have the same imbalance issue (or non-issue) as the current approach?
